src/protocol/boolean/comparison.rs

```diff
@@ -2,10 +2,126 @@ use super::or::or;
 use crate::{
     error::Error,
     ff::Field,
-    protocol::{context::Context, BasicProtocols, BitOpStep, RecordId},
+    protocol::{
+        boolean::{random_bits_generator::RandomBitsGenerator, RandomBits},
+        context::Context,
+        BasicProtocols, BitOpStep, RecordId,
+    },
     secret_sharing::Arithmetic as ArithmeticSecretSharing,
 };
 
+// Adapted from 6.1 Interval Test Protocol in "Multiparty Computation for Interval, Equality, and
+// Comparison Without Bit-Decomposition Protocol", Nishide & Ohta, PKC 2007.
+// <https://doi.org/10.1007/978-3-540-71677-8_23>
+//
+// The version in the paper tests c_1 < x < c_2. For us, c_1 is zero (which eliminates the `c < c_1`
+// case enumerated in the paper), we test ≤ rather than <, and we finally return `x > c_2` which is
+// ~(0 ≤ x ≤ c_2).
+//
+// The remainder of this description names the variables consistently with other routines in this
+// file, which map to the paper as follows:
+//
+// Paper  Ours
+// x      a
+// c      b
+// c_1    0
+// c_2    c
+//
+// Goal: Compute `a > c` as `~(0 ≤ a ≤ c)`
+//
+// Strategy:
+//  1. Generate random r
+//  2. Reveal b = a + r
+//  3. Derive bounds r_low and r_high from public values, such that the desired result is a simple
+//     function of `r_low < r` and `r < r_high`.
+//
+// Case 1: b > c
+// b - c - 1 < r < b + 1
+//  ⟺  b - c ≤ r ≤ b
+//  ⟺  b - b ≤ b - r ≤ b - (b - c)
+//  ⟺  0 ≤ a ≤ c
+//  ⟺  ~(a > c)
+//
+// Case 2: b ≤ c
+//  b < r < b + p - c
+//  ⟺  b - (b + p - c) < b - r < b - b
+//  ⟺  -(p - c) < a < 0
+//  ⟺  a > c
+//
+/// # Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+pub async fn greater_than_constant<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    rbg: &RandomBitsGenerator<F, S, C>,
+    a: &S,
+    c: u128,
```

Comment:
I don't think so. Below this is compared to b.as_u128(). Since "less than" and "more than" aren't really *defined* in a Ring like this - I think it's best to use a u128.

Reply:
in this case we would probably need to enforce `c` < `F::PRIME`?